videostore/indexer/indexer.go

```diff
@@ -0,0 +1,652 @@
+		deleted_at TIMESTAMP
+	);
+	`, segmentsTableName)
+
+	_, err := db.ExecContext(ctx, query)
+	return err
+}
+
+// run starts the indexer event loop.
+func (ix *Indexer) run(ctx context.Context) {
+	ix.logger.Debug("running indexer")
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			ix.logger.Debug("indexer has stopped")
+			return
+		case <-ticker.C:
+			// This is not the initial setting up of the db, but rather to ensure recovery from manual db deletions
+			err := ix.setupDB(ctx)
+			if err != nil {
+				ix.logger.Errorw("error setting up indexer", "error", err)
+				continue
+			}
+			err = ix.refreshIndexAndStorage(ctx)
+			if err != nil {
+				if isDBReadOnlyError(err) {
+					ix.logger.Warnw("readonly DB error detected by indexer refresh, attempting re-setup.", "error", err)
+					ix.dbMu.Lock()
+					if ix.db != nil {
+						dbCloseErr := ix.db.Close()
+						if dbCloseErr != nil {
+							ix.logger.Errorw("error closing readonly db connection", "error", dbCloseErr)
+						}
+						ix.db = nil
+						ix.logger.Debug("readonly database connection closed, will recreate on next tick")
+					}
+					ix.dbMu.Unlock()
+				} else {
+					ix.logger.Errorw("error refreshing indexer", "error", err)
+				}
+			}
+		}
+	}
+}
+
+func isDBReadOnlyError(err error) bool {
+	var sErr sqlite3.Error
+	return errors.As(err, &sErr) && sErr.Code == sqlite3.ErrReadonly
+}
+
+// refreshIndexAndStorage refreshes the index and storage.
+func (ix *Indexer) refreshIndexAndStorage(ctx context.Context) error {
+	ix.dbMu.RLock()
+	defer ix.dbMu.RUnlock()
+	if ix.db == nil {
+		return errors.New("indexer not setup")
+	}
+
+	start := time.Now()
+
+	// 1. Index new files
+	if err := ix.indexNewFiles(ctx); err != nil {
+		return err
+	}
+	ix.logger.Debugf("TIMING: indexNewFiles took %v", time.Since(start))
+
+	// 2. Mark files to delete based on storage limits
+	cleanupDBStart := time.Now()
+	if err := ix.cleanupDB(ctx); err != nil {
+		return err
+	}
+	ix.logger.Debugf("TIMING: cleanupDB took %v", time.Since(cleanupDBStart))
+
+	// 3. Delete marked files from disk and their records from DB
+	cleanupFilesStart := time.Now()
+	if err := ix.cleanupFiles(ctx); err != nil {
+		return err
+	}
+	ix.logger.Debugf("TIMING: cleanupFiles took %v", time.Since(cleanupFilesStart))
+
+	ix.logger.Debugf("TIMING: refreshIndexAndStorage completed in %v", time.Since(start))
+	return nil
+}
+
+// indexNewFiles indexes new video files on disk. Assumes the dbMu is read-locked.
+func (ix *Indexer) indexNewFiles(ctx context.Context) error {
+	indexedFiles, err := ix.getIndexedFiles(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get indexed files: %w", err)
+	}
+
+	diskFileEntries, err := ix.getDiskFilesSorted(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get and sort disk files: %w", err)
+	}
+
+	for _, entry := range diskFileEntries {
+		if ctx.Err() != nil {
+			return ctx.Err()
```

Comment:
aren't we going to log errors here https://github.com/viam-modules/video-store/pull/65/files#diff-5d306141e7eea456775630360bf7be76123bf98072b91bdc846c4f4ce0fce8c8R174 when the context is cancelled (which is a shutdown condition, not an error condition)?